File: API/_archive/GraphQL/python/usecase2_dataloader.py
import asyncio
from typing import List
import strawberry
from strawberry.dataloader import DataLoader

# Simulated batch database query
async def load_posts_for_users(keys: List[str]) -> List[List[str]]:
    print(f"Executing Batch DB Query for keys: {keys}")
    # SELECT * FROM posts WHERE user_id IN keys
    mock_db = {
        "user1": ["Post A", "Post B"],
        "user2": ["Post C"]
    }
    return [mock_db.get(key, []) for key in keys]

post_loader = DataLoader(load_fn=load_posts_for_users)

@strawberry.type
class User:
    id: str
    
    @strawberry.field
    async def posts(self) -> List[str]:
        # Enqueues the request. Loader waits a tick, batches keys, and executes load_posts_for_users
        return await post_loader.load(self.id)

@strawberry.type
class Query:
    @strawberry.field
    def users(self) -> List[User]:
        return [User(id="user1"), User(id="user2")]

schema = strawberry.Schema(query=Query)

async def run_query():
    query = """
        query {
            users {
                id
                posts
            }
        }
    """
    result = await schema.execute(query)
    print("Query Result:", result.data)

if __name__ == "__main__":
    asyncio.run(run_query())

File: API/_archive/GraphQL/golang/usecase2_dataloader.go
package main

import (
	"context"
	"fmt"
	"github.com/graph-gophers/dataloader"
)

// In a real application, this would query a database.
func loadPostsForUsers(ctx context.Context, keys dataloader.Keys) []*dataloader.Result {
	results := make([]*dataloader.Result, len(keys))
	
	// Simulated DB hit for multiple keys in ONE query
	fmt.Printf("Simulating SELECT * FROM posts WHERE user_id IN %v\n", keys)
	
	mockDB := map[string][]string{
		"user1": {"Post A", "Post B"},
		"user2": {"Post C"},
	}

	for i, key := range keys {
		userID := key.String()
		posts := mockDB[userID]
		results[i] = &dataloader.Result{Data: posts, Error: nil}
	}
	
	return results
}

func main() {
	// Initialize the dataloader
	postLoader := dataloader.NewBatchedLoader(loadPostsForUsers)
	
	// Simulate GraphQL resolver concurrently requesting posts for multiple users
	// Instead of 3 DB queries, the dataloader batches them into 1.
	ctx := context.Background()
	thunk1 := postLoader.Load(ctx, dataloader.StringKey("user1"))
	thunk2 := postLoader.Load(ctx, dataloader.StringKey("user2"))
	
	posts1, _ := thunk1()
	posts2, _ := thunk2()
	
	fmt.Printf("User 1 Posts: %v\n", posts1)
	fmt.Printf("User 2 Posts: %v\n", posts2)
}
